src/libtls/tls_peer.c

```diff
@@ -286,6 +286,28 @@ static status_t process_server_hello(private_tls_peer_t *this,
 	}
 	extensions->destroy(extensions);
 
+	/* downgrade protection (see RFC 8446, section 4.1.3) */
+	server_random_end = chunk_create(&this->server_random[24], 8);
+	if (this->tls->get_version_max(this->tls) == TLS_1_3 &&
+		version < TLS_1_3)
+	{
+		if (chunk_equals(server_random_end, tls_downgrade_protection_tls11) ||
+			chunk_equals(server_random_end, tls_downgrade_protection_tls12))
+		{
+			this->alert->add(this->alert, TLS_FATAL, TLS_ILLEGAL_PARAMETER);
```

Comment:
A debug message could be helpful here:
`DBG2(DBG_TLS, "invalid server random possible downgrade attack");`